job-runner/chain/reaper.go

```diff
@@ -558,38 +563,71 @@ func (r *reaper) sendFinalState(finishedAt time.Time) {
 	}
 }
 
-// prepareSequenceRetry prepares a sequence to be retried:
-//   1. Mark failed job and all previously completed jobs in sequence as PENDING.
-//   2. Decrement the number of jobs that were previously run (failed job +
-//   previously completed jobs in sequence) from the chain job count.
-//   3. Increment retry count for the sequence.
-func (r *reaper) prepareSequenceRetry(failedJob proto.Job) {
+// prepareSequenceRetry prepares a sequence to retry. The caller should check
+// r.chain.CanRetrySequence first; this func does not check the seq retry limit
+// or increment seq try count (that's done in traverser.runJobs when the seq
+// start job runs).
+func (r *reaper) prepareSequenceRetry(failedJob proto.Job) proto.Job {
 	sequenceStartJob := r.chain.SequenceStartJob(failedJob.Id)
 
+	seqLogger := r.logger.WithFields(log.Fields{"sequence_id": sequenceStartJob.SequenceId})
+	seqLogger.Info("preparing sequence retry")
+
 	// sequenceJobsToRetry is a list containing the failed job and all previously
 	// completed jobs in the sequence. For example, if job C of A -> B -> C -> D
 	// fails, then A and B are the previously completed jobs and C is the failed
 	// job. So, jobs A, B, and C will be added to sequenceJobsToRetry. D will not be
 	// added because it was never run.
-	var sequenceJobsToRetry []proto.Job
-	sequenceJobsToRetry = append(sequenceJobsToRetry, failedJob)
-	sequenceJobsCompleted := r.sequenceJobsCompleted(sequenceStartJob)
-	sequenceJobsToRetry = append(sequenceJobsToRetry, sequenceJobsCompleted...)
+	sequenceJobsToRetry := r.sequenceJobsCompleted(sequenceStartJob)
+
+	// @todo: fixme: sometimes we have failed job, sometimes we don't
```

Comment:
I'm not sure what needs to be fixed here

Reply:
Updated comment.